src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingWatcher.cs

```diff
@@ -129,93 +108,44 @@ private FileChangeList ComputeChangesAndUpdateState()
             return changes;
         }
 
-        // returns true for empty, '.', and '..' 
-        private unsafe bool IsSpecial(char* cFileName)
-        {
-            if (cFileName[0] == 0) return true;
-            if (cFileName[0] == '.') {
-                if (cFileName[1] == 0) return true;
-                if (cFileName[1] == '.' && cFileName[2] == 0) return true;
-            }
-            return false;
-        }
-
-        private unsafe static int GetLength(char* nullTerminatedString)
-        {
-            int length = -1;
-            while (true)
-            {
-                if (nullTerminatedString[++length] == 0)
-                {
-                    break;
-                }
-            }
-            return length;
-        }
-        private unsafe static bool EndsWith(char* nullTerminatedString, int length, string possibleEnding)
-        {
-            if(possibleEnding.Length > length)
-            {
-                return false;
-            }
-
-            var start = nullTerminatedString + (length - possibleEnding.Length);
-
-            for(int i=0; i<possibleEnding.Length; i++)
-            {
-                if (start[i] != possibleEnding[i]) return false;
-            }
-            return true;
-        }
-
-        private unsafe bool IsWatched(char* filename)
+        internal bool IsWatched(ref FileSystemEntry entry)
         {
+            if (entry.IsDirectory) return false;
             if (_extensionsToWatch == null) return true;
-            var length = GetLength(filename);
-            foreach(var extension in _extensionsToWatch)
+            foreach (var extension in _extensionsToWatch)
             {
-                if(EndsWith(filename, length, extension))
-                {
+                var ignoreCase = !RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
```

Comment:
It might be better to ignore case for Mac & Windows rather than not Linux. Other non-Linux based Unix variants should also be case sensitive. Also a nit- use `bool` and `string`, not var.  The existing code wasn't following our standards regarding var.